%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  
%  KCCG WGS Performance Reporter -- Report generator
%  
%  Usage: 
%    Rscript --vanilla -e "library(knitr); knit('report.Rnw', output = 'report.tex')"
%  
%
%  Mark Pinese, 2015
%  
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\documentclass[10pt,a4paper]{article}
\usepackage{geometry}
\usepackage{amsmath}
\usepackage{booktabs}
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor}
\usepackage{url}

\usepackage{fancyhdr}
\setlength{\headheight}{15.2pt}
\pagestyle{fancyplain}
\usepackage{lastpage}

\begin{document}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% PREPARATION
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% LIBRARIES AND HELPER FUNCTIONS
%--------------------------------------------------------------------
<<load-libs, cache=FALSE, echo=FALSE>>=
suppressMessages(library(VariantAnnotation))
suppressMessages(library(GenomicRanges))
suppressMessages(library(ggplot2))
suppressMessages(library(plyr))
suppressMessages(library(xtable))

suppressMessages(source("report_functions.R"))
@


% LOAD PRECOMPUTED RESULTS
%--------------------------------------------------------------------
<<load-results, cache=FALSE, echo=FALSE>>=
temp = readRDS("report_data.rds")
params = temp$params
class_subsets.performance_thresholded = temp$class_subsets.performance_thresholded
report = temp$report
regions = temp$regions
regions.orig = temp$regions.orig
universe = temp$universe
@


% KNITR SETUP
%--------------------------------------------------------------------
<<setup, cache=FALSE, echo=FALSE>>=
library(knitr)
options(
	tikzDocumentDeclaration = "\\documentclass[11pt]{article}",
	tikzLatexPackages = c(
		getOption("tikzLatexPackages"),
		"\\usepackage{amsmath}"),
	tikzMetricsDictionary="tikzMetrics"
)
opts_chunk$set(
	echo = FALSE, results = 'markup', message = FALSE, warning = FALSE, error = TRUE, 
	fig.align = 'center', cache = FALSE, cache.lazy = FALSE)
opts_knit$set(progress = TRUE, verbose = TRUE)
options(warn = 1)
@


% SUPPLIED REGION BED DETAILS
%--------------------------------------------------------------------
<<region-details, echo=FALSE, cache=FALSE>>=
temp.region_md5 = "NA"
temp.region_label = "NO"
if (params$region.subset)
{
    temp.region_label = "YES"
    temp.region_md5 = fileMD5(params$region.subset.path)
}
@


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% REPORT STARTS HERE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcommand{\HRule}{\rule{\linewidth}{0.5mm}}

\fancyhf{}
\lhead{WGS Performance Report (v\texttt{\Sexpr{params$version$script}}): \texttt{\Sexpr{texquote(params$sample.id)}}}
\rhead{Page \thepage\ of \pageref{LastPage}}

\let\endtitlepage\relax
\begin{titlepage}
\begin{flushleft}
\LARGE{WGS Performance Report: \texttt{\Sexpr{texquote(params$sample.id)}}}
\HRule
\end{flushleft}
\end{titlepage}


\section{Summary}
Sample ID: \texttt{\Sexpr{texquote(params$sample.id)}} \\
Input VCF: \texttt{\url{\Sexpr{params$path.test.orig}}} \\
Analysis restricted to regions? \Sexpr{temp.region_label} \\
Report version: \texttt{\Sexpr{params$version$script}} \\
Report time: \texttt{\Sexpr{texquote(report$gentime)}} \\
Call criterion: \Sexpr{report$criterion_latex} \\


<<subset-calcs>>=
temp.size_genome = sum(as.numeric(width(universe$genome)))      # | GENOME |
temp.size_subset = sum(as.numeric(width(universe$subset)))      # | BED ^ GENOME |
temp.size_gold = sum(as.numeric(width(universe$analysis)))      # | GIAB ^ BED ^ GENOME |
@

\begin{itemize}
\item Of \Sexpr{temp.size_genome} bases in the genome, \Sexpr{temp.size_subset} (\Sexpr{temp.size_subset / temp.size_genome * 100}\%) were in the target analysis regions.
\item Of \Sexpr{temp.size_subset} bases in the target analysis regions, \Sexpr{temp.size_gold} (\Sexpr{temp.size_gold / temp.size_subset * 100}\%) had gold-standard genotype available.
% \item Of \Sexpr{temp.size_gold} target bases with gold-standard calls, \Sexpr{temp.size_gold_lcr} (\Sexpr{temp.size_gold_lcr / temp.size_gold * 100}\%) were in low-complexity regions.
% \item Of \Sexpr{temp.size_notgold} target bases without gold-standard calls, \Sexpr{temp.size_notgold_lcr} (\Sexpr{temp.size_notgold_lcr / temp.size_notgold * 100}\%) were in low-complexity regions.
\end{itemize}

<<subset-plots,fig.height=4,fig.width=6.4,out.width='4in'>>=
plotGenomeBreakdown(f_targ_of_wg = temp.size_subset / temp.size_genome, f_gold_of_targ = temp.size_gold / temp.size_subset)
@


<<performance-plots,fig.width=5,fig.height=5,out.width='2.5in',fig.show='hold'>>=
marginalizePerformance = function(perf_data, subset, vars, ...)
{
    ddply(perf_data[eval(subset, perf_data, parent.frame()),], vars, function(x) { 
        ntp = sum(x$ntp)
        nfn = sum(x$nfn)
        nfp = sum(x$nfp)
        ntn = sum(x$ntn)

        if (ntp + nfn == 0)
        {
            sens = NA
            sens.lci = NA
            sens.uci = NA
        }
        else
        {
            ci_test = binom.test(ntp, ntp + nfn, ...)
            sens = as.vector(ci_test$estimate)
            sens.lci = ci_test$conf.int[1]
            sens.uci = ci_test$conf.int[2]
        }

        c("sens" = sens, "sens.lci" = sens.lci, "sens.uci" = sens.uci, "n" = ntp + nfn + nfp + ntn, ntp = ntp, nfn = nfn, nfp = nfp, ntn = ntn)
    }, .drop = TRUE)
}


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg)), aes(x = zyg, y = sens, fill = zyg)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    geom_errorbar(aes(ymin = sens.lci, ymax = sens.uci), width = 0.2) + 
    labs(x = "", y = "Sensitivity", fill = "Zygosity", title = "SNV Detection") + 
    theme_bw() + ylim(0, 1)

ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg)), aes(x = zyg, y = n, fill = zyg)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    labs(x = "", y = "Number of mutations", fill = "Zygosity", title = "SNV Counts in Gold Standard") + 
    theme_bw()


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg, depth)), aes(x = depth, y = sens, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_ribbon(aes(ymin = sens.lci, ymax = sens.uci, fill = zyg), alpha = 0.25, colour = NA) + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Sequence depth (inclusive)", y = "Sensitivity", fill = "Zygosity", colour = "Zygosity", title = "SNV Detection vs Depth") + 
    theme_bw() + ylim(0, 1)

ggplot(
    marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg, depth)), 
    aes(x = depth, y = ntp + nfn, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Sequence depth (inclusive)", y = "Number of mutations", fill = "Zygosity", colour = "Zygosity", title = "SNV Counts in Gold Standard") + 
    theme_bw()


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Ins" & zyg %in% c("RA", "AA")), .(zyg, mutsize)), aes(x = mutsize, y = sens, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_ribbon(aes(ymin = sens.lci, ymax = sens.uci, fill = zyg), alpha = 0.25, colour = NA) + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases inserted (inclusive)", y = "Sensitivity", fill = "Zygosity", colour = "Zygosity", title = "Insertion Detection Performance") + 
    theme_bw() + ylim(0, 1)

ggplot(
    marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Ins" & zyg %in% c("RA", "AA")), .(zyg, mutsize)), 
    aes(x = mutsize, y = ntp + nfn, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases inserted (inclusive)", y = "Number of mutations", fill = "Zygosity", colour = "Zygosity", title = "Insertion Counts in Gold Standard") + 
    theme_bw()


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Del" & zyg %in% c("RA", "AA")), .(zyg, mutsize)), aes(x = mutsize, y = sens, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_ribbon(aes(ymin = sens.lci, ymax = sens.uci, fill = zyg), alpha = 0.25, colour = NA) + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases deleted (inclusive)", y = "Sensitivity", fill = "Zygosity", colour = "Zygosity", title = "Deletion Detection Performance") + 
    theme_bw() + ylim(0, 1)

ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Del" & zyg %in% c("RA", "AA")), .(zyg, mutsize)), aes(x = mutsize, y = ntp + nfn, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases deleted (inclusive)", y = "Number of mutations", fill = "Zygosity", colour = "Zygosity", title = "Deletion Counts in Gold Standard") + 
    theme_bw()



# Depth FPR: For each depth bin, divide nfp by n.  This is rough as n is the number of variants, not the number of bases,
# in a given depth window.  This approximation assumes that variants are evenly distributed, irrespective of depth -- 
# probably not the case.
temp_nfp = marginalizePerformance(class_subsets.performance_thresholded, quote(TRUE), .(depth))
temp_nfp$approx_fpr = temp_nfp$nfp / (temp_nfp$ntp + temp_nfp$nfn) * sum(temp_nfp$ntp + temp_nfp$nfn) / sum(as.numeric(width(universe$analysis))) * 1e6
ggplot(temp_nfp, aes(x = depth, y = approx_fpr)) + 
    geom_bar(stat = "identity") + 
    labs(x = "Sequence depth (inclusive)", y = "False positives per Mb (approx.)", title = "False positive rate vs depth") + 
    theme_bw()

# temp_nfp = marginalizePerformance(class_subsets.performance_thresholded, quote(TRUE), .(depth, mdust))
# temp_nfp$approx_fpr = temp_nfp$nfp / temp_nfp$n * sum(temp_nfp$n) / sum(as.numeric(width(universe$analysis))) * 1e6
# ggplot(temp_nfp, aes(x = depth, y = approx_fpr, fill = mdust)) + 
#     geom_bar(stat = "identity", position = "dodge") + 
#     labs(x = "Sequence depth (inclusive)", y = "False positives per Mb (approx.)", title = "False positive rate vs depth") + 
#     theme_bw()

ggplot(temp_nfp, aes(x = depth, y = n)) + 
    geom_bar(stat = "identity") + 
    labs(x = "Sequence depth (inclusive)", y = "Variant count at this depth", title = "Variant count vs depth") + 
    theme_bw()



false_positive_count = sum(subset(class_subsets.performance_thresholded, muttype == "None")$nfp)
false_positive_count_subset_size = sum(as.numeric(width(universe$analysis)))
false_positive_rate_per_Mb = false_positive_count / false_positive_count_subset_size * 1e6

ggplot(data.frame(rate_per_Mb = false_positive_rate_per_Mb), aes(x = 1, y = rate_per_Mb)) + 
    geom_bar(stat = "identity") + 
    labs(y = "False positive rate per Mb", title = "False Positive Rate") + 
    theme_bw()
@



% DATA AND SOFTWARE VERSIONS
%--------------------------------------------------------------------
\section{Versions}
\begin{itemize}
\item Software: \begin{itemize}
	\item Performance script: \texttt{\Sexpr{texquote(params$version$script)}}
	\item R: \texttt{\Sexpr{texquote(R.version$version.string)}} (\texttt{\Sexpr{texquote(R.version$platform)}})
	\item Genome: \texttt{\Sexpr{texquote(params$genome)}} (\texttt{\Sexpr{packageVersion(params$genome)}})
	\item Java: \texttt{\Sexpr{texquote(params$version$java)}}
	\item RTG core: \texttt{\Sexpr{texquote(params$version$rtg)}}
	\item Bedtools: \texttt{\Sexpr{texquote(params$version$bedtools)}}
	\item Execution host: \texttt{\Sexpr{texquote(params$version$host)}}
	\item Execution time: \Sexpr{texquote(report$gentime)}
\end{itemize}
\item Data: \begin{itemize}
	\item Input VCF: \texttt{\url{\Sexpr{params$path.test.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.test.orig))}})
	\item Sample ID: \texttt{\Sexpr{texquote(params$sample.id)}}
	\item GiaB VCF: \texttt{\url{\Sexpr{params$path.gold.variants.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.gold.variants.orig))}})
	\item GiaB BED: \texttt{\url{\Sexpr{params$path.gold.regions.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.gold.regions.orig))}})
	\item Analysis restricted to regions? \Sexpr{temp.region_label} \begin{itemize}
		\item Analysis region BED: \texttt{\url{\Sexpr{params$region.subset.path}}} (MD5 \texttt{\Sexpr{texquote(temp.region_md5)}})
	\end{itemize}
	% \item Sequence masks: \texttt{\url{\Sexpr{params$path.mask.regions.prefix}}} \begin{itemize}
		% \item Low complexity: \texttt{\Sexpr{gsub(".*/", "", params$path.mdust)}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.mdust))}})
	% \end{itemize}
\end{itemize}

\end{document}
